stdlib/Markdown/src/render/terminal/render.jl

```diff
@@ -16,118 +16,130 @@ end
     end
 end
 
 function term(io::IO, md::List, columns)
     for (i, point) in enumerate(md.items)
-        print(io, ' '^2margin, isordered(md) ? "$(i + md.ordered - 1). " : "•  ")
-        print_wrapped(io, width = columns-(4margin+2), pre = ' '^(2margin+3),
-                          i = 2margin+2) do io
-            term(io, point, columns - 10)
+        bullet = isordered(md) ? "$(i + md.ordered - 1)." : "• "
+        print(io, ' '^2margin, styled"{markdown_list:$bullet} ")
+        content = annotprint(term, point, columns - 10)
+        lines = split(rstrip(content), '\n')
+        for (l, line) in enumerate(lines)
+            l > 1 && print(io, ' '^(2margin+3))
+            print(io, lstrip(line))
+            l < length(lines) && println(io)
         end
-        i < lastindex(md.items) && print(io, '\n', '\n')
-    end
-end
-
-function _term_header(io::IO, md, char, columns)
-    text = terminline_string(io, md.text)
-    with_output_color(:bold, io) do io
-        pre = ' '^margin
-        print(io, pre)
-        line_no, lastline_width = print_wrapped(io, text,
-                                                width=columns - 4margin; pre)
-        line_width = min(lastline_width, columns)
-        if line_no > 1
-            line_width = max(line_width, div(columns, 3)+length(pre))
-        end
-        header_width = max(0, line_width-length(pre))
-        char != ' ' && header_width > 0 && print(io, '\n', ' '^(margin), char^header_width)
+        i < length(md.items) && print(io, '\n'^(1 + md.loose))
     end
 end
 
 const _header_underlines = collect("≡=–-⋅ ")
 # TODO settle on another option with unicode e.g. "≡=≃–∼⋅" ?
 
-function term(io::IO, md::Header{l}, columns) where l
+function term(io::AnnotIO, md::Header{l}, columns) where l
+    face = Symbol("markdown_h$l")
     underline = _header_underlines[l]
-    _term_header(io, md, underline, columns)
+    pre = ' '^margin
+    local line_width
+    with_output_annotations(io, :face => face) do io
+        headline = annotprint(terminline, md.text)
+        lines = wraplines(headline, columns - 4margin)
+        for (i, line) in enumerate(lines)
+            print(io, pre, line)
+            i < length(lines) && println(io)
+        end
+        line_width = min(textwidth(lines[end]), columns)
+        if length(lines) > 1
+            line_width = max(textwidth(lines[end]), div(columns, 3)+length(pre))
+        end
+    end
+    header_width = max(0, line_width)
+    if underline != ' ' && header_width > 0
+        print(io, '\n', ' '^(margin))
+        with_output_annotations(io -> print(io, underline^header_width), io, :face => face)
+    end
 end
 
 function term(io::IO, md::Code, columns)
-    with_output_color(:cyan, io) do io
-        L = lines(md.code)
-        for i in eachindex(L)
-            print(io, ' '^margin, L[i])
-            i < lastindex(L) && println(io)
-        end
+    code = if md.language ∈ ("", "julia", "julia-repl", "jldoctest")
```

Comment:
```suggestion
    code = if md.language ∈ ("julia", "julia-repl", "jldoctest")
```

just arbitrary \<raw\> blocks marked with \`\`\` shouldn't get julia-formatted, right?

Reply:
I guess it depends how you interpret them. I tend to see them as "whatever the default language is" which on GitHub is `text` and in Julia docstrings is `julia`.

Ultimately, it's a heuristic choice, and I may well have gone the wrong way here.

Reply:
I think raw is the conservative choice here to stick with

Reply:
Sure. It would be nice to syntax highlight function signatures though, which usually appear like so:

```julia
"""
    annotate!(char::AnnotatedChar, label::Symbol => value)

Annotate `char` with the pair `label => value`.
"""
```

Reply:
True. Probably should be a separate PR though, since it is potentially a more significant change that downstream consumers would not be able to distinguish between annotated and un-annotated content

Reply:
> since it is potentially a more significant change that downstream consumers would not be able to distinguish between annotated and un-annotated content

Err, I'm not actually sure what you're talking about here.

Reply:
The StyledString would have julia annotations inside of it instead of marking the section as generic code, no?

Reply:
Yes, the initial code element of the docstring would be highlighted as Julia code, if that's what you mean.

Reply:
@KristofferC might have some helpful comments here, since OhMyREPL does Julia syntax highlighting of markdown code blocks.

Reply:
It seems like we would need a way to distinguish arbitrary Markdown that is being formatted from specifically Julia docstrings that are being rendered

Reply:
The only way to conclusively tell, is to use language annotations on code blocks. Short of that, it's a guess/default. I think "guess Julia because we're in Julia" and requiring `raw`/`text` to be explicit makes sense, but as you demonstrated earlier that's not the only reasonable position.

Reply:
If defaulting to rendering Julia syntax gives us highlighting in the docstrings and otherwise not, then I think we should default to Julia.